### ISSUE TITLE
ultralytics data augmentation functions

### ISSUE THREAD

**Issue by compVisual** [ultralytics/ultralytics#1532]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I know that you can do data augmentation while training a model with the different options explained in the documentation. But I don't know how to, for example rotate my data for diferent degrees; for example 10º, 20º, 40º ... . (applying degrees = 20 allows you to rotate the images 20 degrees).

I want to do several transformations and do a swept from several angles. Is any practical way to do this? For example augmentating the data for 5 degrees each time until reaching a complete circle?

**Comment by N-Friederich** [comment]
@compVisual There is currently no such method to my knowledge. But of course, you can extend the methods in https://github.com/ultralytics/ultralytics/tree/main/ultralytics/yolo/data.

**Comment by compVisual** [comment]
It might be tricky but I'll try

**Comment by glenn-jocher** [comment]
@compVisual I think there may be a misunderstanding here. `degrees: 20` creates a uniform distribution of angles from -20 to + 20, and randomly samples from it for a single image augmentation. If you want a full circle just pass degrees: 180, and images will be rotates across a full circle randomly.

**Comment by N-Friederich** [comment]
@glenn-jocher I thought the goal was to get 0-180 with one step_size at a time. But possible that I misunderstood @compVisual ?

**Comment by glenn-jocher** [comment]
Albumentations also has a random90 option here:
https://albumentations.ai/docs/api_reference/augmentations/geometric/rotate/

**Comment by compVisual** [comment]
@N-Friederich @glenn-jocher yes, the goal is to achive small rotations for example with step_size = 1 degree in order to get a full rotation of each image for example.

**Comment by N-Friederich** [comment]
@compVisual I think this method is very related to your application. In general, you need something like this rather rarely or not at all. Therefore, I think (@glenn-jocher  correct me gladly) that it does not need an implementation here in this repo. But you can easily make a modification, e.g. in this class by adding a new parameter to the __init__ and setting static/class variables.

**Comment by compVisual** [comment]
> @compVisual I think this method is very related to your application. In general, you need something like this rather rarely or not at all. Therefore, I think (@glenn-jocher correct me gladly) that it does not need an implementation here in this repo. But you can easily make a modification, e.g. in this class by adding a new parameter to the **init** and setting static/class variables.

I think that when you have a small data base it will be useful to have a function to rotate the images several times defining a step-size.

**Comment by N-Friederich** [comment]
> Okay, I have a doubt, the rotate option (class albumentations.augmentations.geometric.rotate.Rotate) if you set up the limit to 180 the method will pick each image and rotate it a random angle between (-180, 180) degrees ? I think that what I need is something similar to RandomRotate90 where you can enter a factar to determinate the times you want to rotate the input (even if its randomly) but with an angle different to 90.

@compVisual It depends on exactly how you want to use the method. You could add a parameter step_size to the init. If step_size != -1 (e.g. -1 := random) then the methods should not work randomly, but do rotations, translations or whatever exactly for a specific scalar. For that, you just restrict the list by making the selection,
https://github.com/ultralytics/ultralytics/blob/8c8c43922c8a01f70ef462a3213c270648275bb3/ultralytics/yolo/data/augment.py#L236
 so this that is not done randomly, but for a specific rotation.

**Comment by compVisual** [comment]
> It depends on exactly how you want to use the method. You could add a parameter step_size to the init. If step_size != -1 (e.g. -1 := random) then the methods should not work randomly, but do rotations, translations or whatever exactly for a specific scalar. For that, you just restrict the list by making the selection,

Okay perfect I will try this. How can I apply that RandomPerspective class to my training?? I just write the step_size = x(any number for rotation) the same way that you apply the degrees method?

**Comment by glenn-jocher** [comment]
@compVisual RandomPerspective is already used during training. If you want rotated images simply set a value for degrees, i.e. degrees 90. I'll include some advice from YOLOv5 that also applies here below.

👋 Hello! Thanks for asking about **image augmentation**. YOLOv5 🚀 applies online imagespace and colorspace augmentations in the trainloader (but not the val_loader) to present a new and unique augmented [Mosaic](https://github.com/ultralytics/yolov5/blob/90b7895d652c3bd3d361b2d6e9aee900fd67f5f7/utils/datasets.py#L678-L732) (original image + 3 random images) each time an image is loaded for training. **Images are never presented twice in the same way**.

![YOLOv5 augmentation](https://user-images.githubusercontent.com/26833433/120995721-f3cfed00-c785-11eb-8ee2-b6ef2fa205e8.jpg)

### Augmentation Hyperparameters

The hyperparameters used to define these augmentations are in your hyperparameter file (default `data/hyp.scratch.yaml`) defined when training:
```
python train.py --hyp hyp.scratch-low.yaml
```
https://github.com/ultralytics/yolov5/blob/b94b59e199047aa8bf2cdd4401ae9f5f42b929e6/data/hyps/hyp.scratch-low.yaml#L6-L34

### Augmentation Previews

You can view the effect of your augmentation policy in your train_batch*.jpg images once training starts. These images will be in your train logging directory, typically `yolov5/runs/train/exp`:

`train_batch0.jpg` shows train batch 0 mosaics and labels:  
<img src="https://user-images.githubusercontent.com/26833433/131255960-b536647f-7c61-4f60-bbc5-cb2544d71b2a.jpg" width="800">


### YOLOv5 Albumentations Integration

[YOLOv5](https://github.com/ultralytics/yolov5) 🚀 is now fully integrated with [Albumentations](https://github.com/albumentations-team/albumentations), a popular open-source image augmentation package. Now you can train the world's best Vision AI models even better with custom Albumentations 😃! 

PR https://github.com/ultralytics/yolov5/pull/3882 implements this integration, which will automatically apply Albumentations transforms during YOLOv5 training if `albumentations>=1.0.3` is installed in your environment. See https://github.com/ultralytics/yolov5/pull/3882 for full details.

Example `train_batch0.jpg` on COCO128 dataset with Blur, MedianBlur and ToGray. See the YOLOv5 [Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb) to reproduce: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

<img src="https://user-images.githubusercontent.com/26833433/124400879-ff331b80-dd25-11eb-9b67-fe85ac4ca104.jpg" width="800">

Good luck 🍀 and let us know if you have any other questions!

**Comment by akashAD98** [comment]
@glenn-jocher I have less count for training images, so doing external data augmentations will improve my performace? 
which is better external data augmentation vs  default data augmentations while model is trainig

**Comment by glenn-jocher** [comment]
Hi! Yes, external data augmentation can certainly help improve model performance, especially when you have a relatively small amount of training data. External data augmentation adds more variety to your data set, and can help your model learn to generalize better on unseen data. It's important to note, however, that external data augmentation should still be used in conjunction with default data augmentations (e.g. random crop, random flip, etc.) during training, as these augmentations can help to prevent overfitting and improve model robustness.

That being said, it's important to use external data augmentations that are relevant to your particular use case. For example, if you are working with object detection in autonomous driving, you may want to augment your data with various lighting and weather conditions that may be present in real-world scenarios. 

In summary, both external and default data augmentations can be beneficial for model training. External data augmentations can add more variety to your data set and help with generalization, while default data augmentations can help prevent overfitting and improve model robustness. If you do decide to use external data augmentation, make sure to select augmentations that are relevant to your specific use case. For more information on data augmentation, you can refer to the Ultralytics YOLOv8 documentation at https://docs.ultralytics.com.
